mii/config.py

```diff
@@ -99,29 +153,124 @@ def meta_tensor_or_sys_mem(cls, values):
             )
         return values
 
-    class Config:
-        validate_all = True
-        validate_assignment = True
-        use_enum_values = True
-        extra = 'forbid'
-        json_encoders = {torch.dtype: lambda x: str(x)}
+    @root_validator
+    def zero_dtype_valid(cls, values):
+        if values.get("enable_zero"):
+            if values.get("ds_config").get("fp16", {}).get("enabled", False):
+                assert (
+                    values.get("dtype") == DtypeEnum.float16
+                ), "ZeRO FP16 enabled, `dtype` must be set to `torch.float16`"
+            else:
+                assert (
+                    values.get("dtype") == DtypeEnum.float32
+                ), "ZeRO FP16 disabled, `dtype` must be set to `torch.float32`"
+        return values
 
+    @root_validator
+    def deepspeed_or_zero(cls, values):
+        assert not (
+            values.get("enable_deepspeed") and values.get("enable_zero")
+        ), "DeepSpeed and ZeRO cannot both be enabled, select only one"
+        return values
 
-class ReplicaConfig(BaseModel):
-    hostname: str = ""
-    tensor_parallel_ports: List[int] = []
-    torch_dist_port: int = None
-    gpu_indices: List[int] = []
 
-    class Config:
-        validate_all = True
-        validate_assignment = True
+class MIIConfig(DeepSpeedConfigModel):
+    deployment_config: DeploymentConfig
+    deployment_type: DeploymentType = DeploymentType.LOCAL
+    hf_auth_token: str = ""
+    port_number: int = 50050
+    enable_restful_api: bool = False
+    restful_api_port: int = 51080
+    hostfile: str = DLTS_HOSTFILE
+    version: int = 1
 
+    @root_validator(skip_on_failure=True)
+    def propagate_hf_auth(cls, values):
+        # This validator is for when we support multiple models in a deployment
+        hf_auth_token = values.get("hf_auth_token")
+        deployment_config = values.get("deployment_config")
+        if not deployment_config.hf_auth_token:
+            deployment_config.hf_auth_token = hf_auth_token
+        return values
 
-class LoadBalancerConfig(BaseModel):
-    port: int = None
-    replica_configs: List[ReplicaConfig] = []
+    @root_validator(skip_on_failure=True)
+    def AML_name_valid(cls, values):
+        if values.get("deployment_type") == DeploymentType.AML:
+            allowed_chars = set(string.ascii_lowercase + string.ascii_uppercaes +
+                                string.digits + "-")
+            assert (
+                set(values.get("deployment_config").deployment_name) <= allowed_chars
+            ), "AML deployment names can only contain a-z, A-Z, 0-9, and '-'."
+        return values
 
-    class Config:
-        validate_all = True
-        validate_assignment = True
+    @root_validator(skip_on_failure=True)
+    def generate_replica_configs(cls, values):
```

Comment:
This validator is not great. Need to find a cleaner way to setup the replica configs. Also, we may not want to do replica config construction when we call to `mii.deploy()`. In the case that a user is not spinning up MII on the machine they run `deploy()` (e.g., with AML deployment), then this will cause errors because it may not find the correct number of GPUs :(